File: leet_code/js/countingBits.js
/**
 * @param {number} num
 * @return {number[]}
 */
var countBits = function (num) {
  const res = [0];

  if (num === 0) return res;

  for (let i = 0; i <= num; i++) {
    const np2 = 2 ** Math.floor(Math.log2(i));
    res.push(res[i - np2] + 1);
  }

  return res;
};

File: leet_code/c++/may_challenge/counting_bits.cpp
#include <cmath>
#include <vector>

using namespace std;

class Solution {
 public:
  vector<int> countBits(int num) {
    vector<int> res = {0};

    if (num == 0)
      return res;

    for (int i = 1; i <= num; i++) {
      int np2 = pow(2, floor(log2(i)));
      res.push_back(res[i - np2] + 1);
    }

    return res;
  }
};